### PR TITLE
Create qcow2 image for all providers

### PR DESCRIPTION
**Pull request by bpradipt**
This PR adds support for creating qcow2 image for all providers. Further it introduces container builds for Pod VM image.
The builder image concept is taken from the PR by @wainersm

### CODE REVIEW

**Comment by bpradipt** [comment]
@jtumber-ibm @snir911 ptal and share your comments.

**Comment by bpradipt** [comment]
For container builds, I tested it by running the following command:
```
docker build -t podvm_builder --build-arg CAA_SRC="https://github.com/bpradipt/cloud-api-adaptor.git" --build-arg CAA_SRC_BRANCH=builder -f Dockerfile.podvm_builder  .
docker build -t podvm_libvirt --build-arg BUILDER_IMG=localhost/podvm_builder:latest -f Dockerfile.podvm .
```

For non-container builds, I tested it by running the following command:
```
cd podvm
CLOUD_PROVIDER=libvirt make image
```

**Review by snir911 — APPROVED** [review]
nit: the instructions in first commit msg are wrong, however, it is correct in the following commit msg so i guess that's fine. thanks! :)

**Review by snir911 — APPROVED** [review]
LGTM

File: podvm/Makefile [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+.PHONY: image binaries clean
+
+UBUNTU_IMAGE_URL  := https://cloud-images.ubuntu.com/$(UBUNTU_RELEASE)/current/$(UBUNTU_RELEASE)-server-cloudimg-$(ARCH).img
+UBUNTU_IMAGE_CHECKSUM := $(shell curl -LO  https://cloud-images.ubuntu.com/"$(UBUNTU_RELEASE)"/current/SHA256SUMS && \
+				grep "$(UBUNTU_RELEASE)"-server-cloudimg-"$(ARCH)".img SHA256SUMS | awk '{print $$1}' && \
+				rm -f SHA256SUMS)
+
+IMAGE_SUFFIX := .qcow2
+PODVM_DISTRO := ubuntu
```
Comment (jtumber-ibm) [inline podvm/Makefile:15]: Should this be `PODVM_DISTRO ?= ubuntu`?

File: podvm/Makefile [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	$(eval OPTS := -var qemu_image_name=${IMAGE_FILE} \
+	-var cloud_image_url=${RHEL_IMAGE_URL} \
+	-var disk_size=11144 \
+	-var ssh_username=cloud-user \
+	-var cloud_image_checksum=${RHEL_IMAGE_CHECKSUM} qcow2/rhel)
+else
+	$(error PODVM_DISTRO is invalid or not defined)
+endif
+
+binaries: $(BINARIES)
```
Comment (jtumber-ibm) [inline podvm/Makefile:44]: Want to include this target in the common makefile?

File: podvm/Dockerfile.podvm [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Builds pod vm image inside container
+#
+ARG BUILDER_IMG="quay.io/confidential-containers/cloud-api-adaptor-podvm-builder"
+
+FROM ${BUILDER_IMG} AS podvm_builder
+
+ARG CLOUD_PROVIDER=libvirt
```
Comment (jtumber-ibm) [inline podvm/Dockerfile.podvm:11]: Should we introduce a `generic` or `null` provider for this image creation rather than use `libvirt`?

File: podvm/Makefile (hunk 1 above)
Comment (bpradipt) [reply podvm/Makefile:15]: yes.

File: podvm/Makefile (hunk 2 above)
Comment (bpradipt) [reply podvm/Makefile:44]: Good suggestion. I'll do it in the common makefile

File: podvm/Dockerfile.podvm (hunk 3 above)
Comment (bpradipt) [reply podvm/Dockerfile.podvm:11]: We don't have any generic or null provider and that needs to be added first.  Let's keep `libvirt` as default now and we can change it once null or generic provider code is added. 
Does it sound reasonable ?

**Review by snir911 — APPROVED** [review]

**Review by jtumber-ibm — APPROVED** [review]

**Comment by wainersm** [comment]
Hi @bpradipt , thanks for making the docker build for the podvm generic.

I got one question before start my review: why you chose to get the peer pods source from github instead of the current directory? It makes the build process a little bit more complex because we will need to push changes to a temporary tree/branch on github.

**Comment by bpradipt** [comment]
> Hi @bpradipt , thanks for making the docker build for the podvm generic.
> 
> I got one question before start my review: why you chose to get the peer pods source from github instead of the current directory? It makes the build process a little bit more complex because we will need to push changes to a temporary tree/branch on github.

Following are the reasons
1. Have the image build process completely de-coupled from any host interaction w.r.to source code. Only requirement is to have a docker or podman runtime. This also enables the pod vm image to be created using a secure build pipeline. 
2. The podvm image creation process is time consuming.  So for every release we can publish the builder image separately from the release branch/commit that can be used by anyone to build their custom pod VM image.
3. I don't expect the pod vm image to be created for testing every PR. It won't scale unless we figure out a time optimal mechanism. The container build mechanism is provided to end users to build their custom pod VM image or as part of our release process to publish qcow2 images.

As for local development workflow, as you mentioned the developer need to use their own github fork/branch to create the builder image followed by the pod vm. However a developer can also override the existing source code in the published builder image by their local copy by using something like this. 
```
FROM _published_builder_image_
ADD .  /src/cloud-api-adaptor
```

**Comment by bpradipt** [comment]
@stevenhorsman @wainersm can we proceed with this PR? Post merging this I want to start working on adding a github action to create qcow2 images as release artifacts

**Comment by stevenhorsman** [comment]
> @stevenhorsman @wainersm can we proceed with this PR? Post merging this I want to start working on adding a github action to create qcow2 images as release artifacts

Yep - if @jtumber-ibm is happy then I am :)

**Review by stevenhorsman — DISMISSED** [review]
Approved on behalf of @jtumber-ibm

**Comment by wainersm** [comment]
Hi @bpradipt !

I gave it a try; worked out but I had to change the Dockerfile as my peer pods tree is not named "cloud-api-adaptor":

```
diff --git a/podvm/Dockerfile.podvm_builder b/podvm/Dockerfile.podvm_builder
index 9680901..e1bc571 100644
--- a/podvm/Dockerfile.podvm_builder
+++ b/podvm/Dockerfile.podvm_builder
@@ -50,8 +50,8 @@ RUN echo $CAA_SRC
 
 RUN echo $CAA_SRC_BRANCH
 
-RUN git clone ${CAA_SRC} -b ${CAA_SRC_BRANCH}
-RUN git clone ${KATA_SRC} -b ${KATA_SRC_BRANCH}
+RUN git clone ${CAA_SRC} -b ${CAA_SRC_BRANCH} cloud-api-adaptor
+RUN git clone ${KATA_SRC} -b ${KATA_SRC_BRANCH} kata-containers
 
 
 ENV GOPATH /src
```

Could you commit that change?

**Comment by bpradipt** [comment]
> > Hi @bpradipt !
> > I gave it a try; worked out but I had to change the Dockerfile as my peer pods tree is not named "cloud-api-adaptor":
> > ```
> > diff --git a/podvm/Dockerfile.podvm_builder b/podvm/Dockerfile.podvm_builder
> > index 9680901..e1bc571 100644
> > --- a/podvm/Dockerfile.podvm_builder
> > +++ b/podvm/Dockerfile.podvm_builder
> > @@ -50,8 +50,8 @@ RUN echo $CAA_SRC
> >  
> >  RUN echo $CAA_SRC_BRANCH
> >  
> > -RUN git clone ${CAA_SRC} -b ${CAA_SRC_BRANCH}
> > -RUN git clone ${KATA_SRC} -b ${KATA_SRC_BRANCH}
> > +RUN git clone ${CAA_SRC} -b ${CAA_SRC_BRANCH} cloud-api-adaptor
> > +RUN git clone ${KATA_SRC} -b ${KATA_SRC_BRANCH} kata-containers
> >  
> >  
> >  ENV GOPATH /src
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you commit that change?
> 
> Sorry, didn't understand.. what do you want me to change ?

Oh I understand :-) ..   But question is why it's needed ?  Can't you override these in your custom Dockerfile ? 
Just to be clear, the Dockerfile I shared was to explain the possibility. You can customise it as needed.

**Comment by bpradipt** [comment]
@wainersm an example Dockerfile for your specific use case of using a different tree
```
FROM _published_builder_image_
RUN rm -fr /src/cloud-api-adaptor && mkdir -p /src/cloud-api-adaptor
ADD .  /src/cloud-api-adaptor
```

**Comment by wainersm** [comment]
> > > Hi @bpradipt !
> > > I gave it a try; worked out but I had to change the Dockerfile as my peer pods tree is not named "cloud-api-adaptor":
> > > ```
> > > diff --git a/podvm/Dockerfile.podvm_builder b/podvm/Dockerfile.podvm_builder
> > > index 9680901..e1bc571 100644
> > > --- a/podvm/Dockerfile.podvm_builder
> > > +++ b/podvm/Dockerfile.podvm_builder
> > > @@ -50,8 +50,8 @@ RUN echo $CAA_SRC
> > >  
> > >  RUN echo $CAA_SRC_BRANCH
> > >  
> > > -RUN git clone ${CAA_SRC} -b ${CAA_SRC_BRANCH}
> > > -RUN git clone ${KATA_SRC} -b ${KATA_SRC_BRANCH}
> > > +RUN git clone ${CAA_SRC} -b ${CAA_SRC_BRANCH} cloud-api-adaptor
> > > +RUN git clone ${KATA_SRC} -b ${KATA_SRC_BRANCH} kata-containers
> > >  
> > >  
> > >  ENV GOPATH /src
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Could you commit that change?
> > 
> > 
> > Sorry, didn't understand.. what do you want me to change ?
> 
> Oh I understand :-) .. But question is why it's needed ? Can't you override these in your custom Dockerfile ? Just to be clear, the Dockerfile I shared was to explain the possibility. You can customise it as needed.

My tree is `https://github.com/wainersm/cc-cloud-api-adaptor`. I ran like below with the change I proposed:

```
docker build -t podvm_builder --build-arg CAA_SRC=https://github.com/wainersm/cc-cloud-api-adaptor --build-arg CAA_SRC_BRANCH=podvm_builder -f Dockerfile.podvm_builder .
```

File: podvm/Dockerfile.podvm [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+# Building binaries separately allows for reuse in case of error
+# with image building step
+RUN cd cloud-api-adaptor/podvm && \
+     CLOUD_PROVIDER=$CLOUD_PROVIDER make binaries
+
+RUN cd cloud-api-adaptor/podvm && \
+     CLOUD_PROVIDER=$CLOUD_PROVIDER make image
```
[the remark this answers is not in the document]
Comment (bpradipt) [reply podvm/Dockerfile.podvm:19]: @whitebrandy let's handle the rhel (including centos image) generation using container builds in a separate PR for the following reasons:
1. We need to figure out a mechanism which avoids copying any files from local host directory and I think it will need some discussions.
2. Use a RHEL or CentOS based builder image for building the dynamic linked binaries or installing the native packages

File: podvm/qcow2/rhel/qemu-rhel.pkr.hcl [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  provisioner "shell" {
+    remote_folder = "~"
+    inline = [
+      "sudo bash ~/misc-settings.sh"
+    ]
+  }
+
+  provisioner "file" {
+    source      = "rhel/selinux_relabel.sh"
```
[the remark this answers is not in the document]
Comment (bpradipt) [reply podvm/qcow2/rhel/qemu-rhel.pkr.hcl:64]: Thanks.

**Comment by bpradipt** [comment]
@stevenhorsman @jtumber-ibm @wainersm @snir911 @whitebrandy ptal
I have addressed all the new suggestions except the RHEL build changes as suggested by @whitebrandy. Explained the reasons  in the following comment - https://github.com/confidential-containers/cloud-api-adaptor/pull/396#discussion_r1042957818

**Comment by bpradipt** [comment]
> This is a very good idea. @snir911 had a similar request to me.

Ref: https://github.com/confidential-containers/cloud-api-adaptor/issues/372
Once this is merged, the next item is to create a github action and add the qcow2 images as release artifacts.

**Comment by bpradipt** [comment]
> Regarding "I don't expect the pod vm image to be created for testing every PR" I agree that we shouldn't rebuild most of the components (e.g. kata-agent). But how do we know that agent-protocol-forwarder shouldn't be rebuilt?

We need to add test cases for providers. Also figure out for e2e tests what's the bare minimum that is needed. Also we need to explore ways to just add the changed bits in the pod vm image for testing instead of rebuilding everything. All in all this will need some thoughts and work.  We'll get there..

**Review by whitebrandy — APPROVED** [review]
LGTM

**Review by jtumber-ibm — APPROVED** [review]
lgtm @stevenhorsman
